core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigInstantiator.java

```diff
@@ -33,18 +34,16 @@
  * has failed and we are attempting to do some form of recovery via hot deployment
  * <p>
  * TODO: fully implement this as required, at the moment this is mostly to read the HTTP config when startup fails
+ * or for basic logging setup in non-Quarkus tests
  */
 public class ConfigInstantiator {
 
     // certain well-known classname suffixes that we support
-    private static Set<String> supportedClassNameSuffix;
+    private static Set<String> SUPPORTED_CLASS_NAME_SUFFIXES = Set.of("Config", "Configuration");
 
-    static {
-        final Set<String> suffixes = new HashSet<>();
-        suffixes.add("Config");
-        suffixes.add("Configuration");
-        supportedClassNameSuffix = Collections.unmodifiableSet(suffixes);
-    }
+    private static final String QUARKUS_PROPERTY_PREFIX = "quarkus.";
+
+    private static final Pattern SEGMENT_EXTRACTION_PATTERN = Pattern.compile("(\"[^\"]+\"|[^.\"]+).*");
```

Comment:
This gets:
- `foo` from `quarkus.foo.bar.baz`
- `"foo.bar"` from `quarkus."foo.bar".baz`

(see also the test I've added)